### DISCUSSION TITLE
How to get the search on a poriton of the release name and extract the hash?

### DISCUSSION THREAD

**Discussion by toast-gear** [renovatebot/renovate#12356] [Request Help]
I've tried but this software is too confusing for me. 

I want to take the release name and look for it in a series of files (really all `*.yaml|*.yml` files in `.github/workflows/*` but I can make multiple regexManagers for that if mutiple files can't be done in a singel regexManager.

I've figured out how to grab the version and do a find and replace (see below) but my releases will be hashes, further more there will be a range of hashes in a workflow and I want a specific hash for a specific action.

```
{
  "extends": ["config:base"],
  "labels": ["dependencies"],
  "packageRules": [
    {
      "matchPackageNames": ["owner/source"],
      "extractVersion": "^v(?<version>.*)$",
      "automerge": false
    }
  ],
  "regexManagers": [
    {
      // use https://github.com/actions/runner/releases
      "fileMatch": [".github/workflows/ci.yml"],
      "matchStrings": ["VERSION: +(?<currentValue>.*?)\\n"],
      "depNameTemplate": "owner/source",
      "datasourceTemplate": "github-releases"
    }
  ]
}
```

My releases will be named things like `say8425/aws-secrets-manager-actions@a561231232f5b31221216760facf7c94c944033b`. What I technically want is for renovate to find instances of `say8425/aws-secrets-manager-actions@` in the `fileMatch` files and replace the hash with the new hash as defined in the release title e.g `e5669f5b3be75fa481216760facf7c94c944033b`.

I'm basically looking to use renovate to bump actions in workflows. We wish to whitelist actions at a commit level for actions from untrusted sources and want a way to automate bumping them for developers. I was thinking I could do this with renovate and github releases. I could setup a matchStrings for each action but I'd quite like to get it from the release name so teh config can be generic and added to a template repository.

**Comment by viceice** [comment]
`["VERSION: +(?<depName>.+?)@(?<currentDigest>.+?)\\s"]`

Renovate isn't able to sort hashes as versions. Can you describe your sample some more detailed.

**Reply by toast-gear** [reply]
I was hoping I could just extract an arbitrary string, I know this is a bit of an edge case so I didn't expect native support. 

You can whitelist what actions are and are not allowed to be ran in a workflow, this can be done at a repository, organisation or enterprise level. you can also whitelist the individual action at an owner, repository, tag or commit level.

![image](https://user-images.githubusercontent.com/15716903/139064437-fc83d1a5-b40b-4f74-827f-42c7b0f9cd59.png)

This is neat however it does present a challenge in terms of management. You don't want to end up with dozens of commits whitelisted on each action. You want to whitelist one, add a second and then update your workflows to use the new hash. I wish to automate that last part of the process and I was thinking of using renovate to do this as the problem is a standard dependency management issue.

My idea was you want to whitelist a new commit hash, e.g. you already have:

```
say8425/aws-secrets-manager-actions@a561231232f5b31221216760facf7c94c944033b
```

you wish to have your workflows bumped to a new versionn, so you whitelist the new hash too:

```
say8425/aws-secrets-manager-actions@a561231232f5b31221216760facf7c94c944033b
say8425/aws-secrets-manager-actions@34nkl32n432nb54o3lb5o435o43ug5o43ug5uo
```

Now you want your workflows to auto raise a PR for the new hash as you have hundreds of repos and you don't want this to be a manual process. My idea was that you publish a new release titled:

```
say8425/aws-secrets-manager-actions@34nkl32n432nb54o3lb5o435o43ug5o43ug5uo
```

Then each repos renovate's config detects the new release in a sort of central dummy repository and does a find and replace for the action. The search pattern being the release title sans the hash `say8425/aws-secrets-manager-actions@` and the replace pattern being on the hash value in the release title `34nkl32n432nb54o3lb5o435o43ug5o43ug5uo`.

I know it's a little hacky but I thought renovate could handle it, whilst the requirement is a bit niche, what is actually required mechnically is just a regex find and replace driven from a GitHub release title. I wanted the renovate config to be generic so the find and replace is driven from the release title instead of config so we can easily push it out everywhere and the config supports all actions without changes needed to it.

**Comment by rarkins** [comment]
I've tried but this question is too confusing for me.

**Reply by toast-gear** [reply]
@rarkins does my reply to @viceic clarify things?

**Reply by toast-gear** [reply]
OK, I'll try to explain further. If you want to lock what action versions can and can't be used in a workflow at the hash level you have to use the whitelist feature (see screenshot). The problem this brings is how do you then automate a PR being raised to bump your actions to use the new whitelisted hash? The whitelist feature doesn't produce a webhook event or anything and it's an internal setting so the only way of publishing the fact that a new version of a whitelisted action is whitelisted is by doing something like creating a release. We can use renoate to ingest github release events so we now have our trigger. The next problem is how do we update the right action reference in a workflow and how do we update it to use the new version? I thought we could do this simply through the github release name which renovate can access

@rarkins Is there anything specific I can try to help clarify?

**Reply by rarkins** [reply]
I think if the software is too confusing for you then you should consider other software.

**Reply by viceice** [reply]
I think you should let open renovate a pr to update the workflow files to new digest as normal, then the ci run should fail, as it's not whitelisted.

So you can check the pr and whitlist the digest (hash) and rerun the ci, now you should be able to merge, or even let renovtae automerge.

**Reply by toast-gear** [reply]
As a more practical example, https://github.com/actions-runner-controller/actions-runner-controller/blob/master/.github/renovate.json5 here is a project which takes the title of a release from another repository, in this case the title is `vX.X.X`, and uses it to replace a the old string.

I just want to take this idea 1 step further and instead of extracting a version, I want to split the release title into 2 parts:
1. Extract the title i.e. everything before the `@`
2. Extract the hash i.e. everything after `@`

then use the former as the find and the latter as the replace

**Reply by viceice** [reply]
you can checkout our workflows, we have pinned all our actions, but we have a renovate comment, so renovate can follow specific github releases and shows us only hashes of the versions. we also see new pr as digest update, if a action is moving the release to a new git tag.
Then we decide to merge or not.

**Reply by toast-gear** [reply]
is there a method for extracting all text before a delimeter, `@` in my case, and all text after a delimeter within a github release title? I think `extractVersion` is just a generic regex search method so I can use that for the hash. I'm not sure though if I can also extract the title (everything before `@`) in the same packageRules and use it as my find?

**Reply by viceice** [reply]
sure, it's a regex, see docs for named capture groups
https://docs.renovatebot.com/modules/manager/regex/

**Comment by HonkingGoose** [comment]
I'm reading your post, and your process is really confusing to me as well.
I think you're both complicating things, and not really gaining any benefit from doing so.

## Fork actions into a trusted organization, cut your own releases

I think maybe you'd benefit from following a different process altogether.

1. Fork any actions that you want to use across multiple repositories in your organization/account
1. Review code that's now in your fork
1. Make a new release from within your fork (can use GitHub auto-generated release notes when making the release)
1. Change references to upstream actions to use the forked action instead
1. Whitelist actions coming from the forked repository in the GitHub Actions permission screen/setting with a wildcard like `forked-actions-organization/*`
1. Let Renovate update your commit hashes for the workflow files [^renovate] by using:
    ```json
    {
      "extends": ["helpers:pinGitHubActionDigests"]
    }
    ```
    - Renovate will add a comment with the "tagged version" of the release, so you have a human readable reference for the version you're using
1. I think that's all you need.

GitHub Universe has released a good video on working with Actions in a secure way [^youtube], I recommend you watch the linked video _before_ you start working on the items listed above.

Review my suggestion for suitability for your organization and process before starting, maybe I'm totally honking up the wrong tree here...

### Benefits

- Full control over the actions you run in your organization
- Forced to do manual review of the code when updating the action via the fork
- You can whitelist/trust the actions from your fork, as you've done the review before cutting the release
- Action review work is now centralized in one place, instead of spread out over multiple repositories

### Drawbacks

- Needs manual migration, which will take a while
- More manual work to review (you should review anyways...)
- Manual work to cut a release (can automate this, or just use the auto-generated changelog within GitHub Release feature)
- If you don't update the fork, you're probably worse off than if you just use the upstream actions as-is

[^youtube]: https://www.youtube.com/watch?v=Ers-LcA7Nmc
[^renovate]: https://docs.renovatebot.com/modules/manager/github-actions/

**Reply by toast-gear** [reply]
@HonkingGoose this sounds like a much better way of doing what I want to do. I'll have a watch of that video and report back so this discussion can be marked as solved. Thanks for going through my rambling dude.
